internal/state/store.go

```diff
@@ -2,84 +2,280 @@ package state
+func newMultiObjectStore(stores map[schema.GroupVersionKind]objectStore) *multiObjectStore {
+	return &multiObjectStore{
+		stores: stores,
 	}
+}
+
+func (m *multiObjectStore) mustFindStoreForObj(obj client.Object) objectStore {
+	objGVK := mustGVKForObject(obj)
 
-	// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-	if s.gc != nil && s.gc.Generation == gc.Generation {
-		resourceChanged = false
+	store, exist := m.stores[objGVK]
+	if !exist {
+		panic(fmt.Errorf("object store for %T %s not found", obj, client.ObjectKeyFromObject(obj)))
 	}
 
-	s.gc = gc
+	return store
+}
+
+func (m *multiObjectStore) get(objType client.Object, nsname types.NamespacedName) (client.Object, bool) {
+	return m.mustFindStoreForObj(objType).get(nsname)
+}
+
+func (m *multiObjectStore) upsert(obj client.Object) {
+	m.mustFindStoreForObj(obj).upsert(obj)
+}
+
+func (m *multiObjectStore) delete(objType client.Object, nsname types.NamespacedName) {
+	m.mustFindStoreForObj(objType).delete(nsname)
+}
+
+type changeTrackingUpdaterObjectTypeCfg struct {
+	gvk                  schema.GroupVersionKind
+	store                objectStore
+	changingUpsertDelete bool
+}
+
+type changeTrackingUpdater struct {
+	store *multiObjectStore
+
+	supportedGKVs            gvkList
+	changingUpsertDeleteGKVs gvkList
+	persistedGKVs            gvkList
+
+	gcName string
 
-	s.changed = s.changed || resourceChanged
+	capturer relationship.Capturer
+	changed  bool
 }
 
-func (s *store) captureGatewayChange(gw *v1beta1.Gateway) {
-	resourceChanged := true
+func newChangeTrackingUpdater(
+	gcName string,
+	capturer relationship.Capturer,
+	objectTypeCfgs []changeTrackingUpdaterObjectTypeCfg,
+) *changeTrackingUpdater {
+	var (
+		supportedGKVs            gvkList
+		changingUpsertDeleteGKVs gvkList
+		persistedGKVs            gvkList
 
-	// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-	prev, exist := s.gateways[client.ObjectKeyFromObject(gw)]
-	if exist && gw.Generation == prev.Generation {
-		resourceChanged = false
+		stores = make(map[schema.GroupVersionKind]objectStore)
+	)
+
+	for _, cfg := range objectTypeCfgs {
+		supportedGKVs = append(supportedGKVs, cfg.gvk)
+
+		if cfg.changingUpsertDelete {
+			changingUpsertDeleteGKVs = append(changingUpsertDeleteGKVs, cfg.gvk)
+		}
+
+		if cfg.store != nil {
+			persistedGKVs = append(persistedGKVs, cfg.gvk)
+			stores[cfg.gvk] = cfg.store
+		}
 	}
 
-	s.gateways[client.ObjectKeyFromObject(gw)] = gw
+	return &changeTrackingUpdater{
+		store:                    newMultiObjectStore(stores),
+		supportedGKVs:            supportedGKVs,
+		changingUpsertDeleteGKVs: changingUpsertDeleteGKVs,
+		persistedGKVs:            persistedGKVs,
+		capturer:                 capturer,
+		gcName:                   gcName,
+	}
+}
 
-	s.changed = s.changed || resourceChanged
+func (s *changeTrackingUpdater) assertSupportedGVK(gvk schema.GroupVersionKind) {
```

Comment:
why do we need to assert this?

Reply:
this is to prevent the client (Updater interface) from upserting unsupported kinds